src/contractor/contractor.cpp

```diff
@@ -45,37 +49,162 @@
     {
-        return static_cast<uint64_t>(k.first) ^ (static_cast<uint64_t>(k.second) << 12);
     }
+
+    bool operator<(const Segment &rhs) const
+    {
+        return std::tie(from, to) < std::tie(rhs.from, rhs.to);
+    }
+    bool operator==(const Segment &rhs) const
+    {
+        return std::tie(from, to) == std::tie(rhs.from, rhs.to);
+    }
+};
+
+struct SpeedSource final
+{
+    unsigned speed;
+    std::uint8_t source;
 };
 
-template <> struct hash<std::tuple<OSMNodeID, OSMNodeID, OSMNodeID>>
+struct Turn final
 {
-    std::size_t operator()(const std::tuple<OSMNodeID, OSMNodeID, OSMNodeID> &k) const noexcept
+    std::uint64_t from, via, to;
+    Turn() : from(0), via(0), to(0) {}
+    Turn(const std::uint64_t &from, const std::uint64_t &via, const std::uint64_t &to)
+        : from(from), via(via), to(to)
+    {
+    }
+    Turn(const OSMNodeID &from, const OSMNodeID &via, const OSMNodeID &to)
+        : from(static_cast<std::uint64_t>(from)), via(static_cast<std::uint64_t>(via)),
+          to(static_cast<std::uint64_t>(to))
+    {
+    }
+    bool operator<(const Turn &rhs) const
+    {
+        return std::tie(from, via, to) < std::tie(rhs.from, rhs.via, rhs.to);
+    }
+    bool operator==(const Turn &rhs) const
     {
-        std::size_t seed = 0;
-        boost::hash_combine(seed, static_cast<uint64_t>(std::get<0>(k)));
-        boost::hash_combine(seed, static_cast<uint64_t>(std::get<1>(k)));
-        boost::hash_combine(seed, static_cast<uint64_t>(std::get<2>(k)));
-        return seed;
+        return std::tie(from, via, to) == std::tie(rhs.from, rhs.via, rhs.to);
     }
 };
-}
 
-namespace
+struct PenaltySource final
 {
+    double penalty;
+    std::uint8_t source;
+};
+
 template <typename T> inline bool is_aligned(const void *pointer)
 {
     static_assert(sizeof(T) % alignof(T) == 0, "pointer can not be used as an array pointer");
     return reinterpret_cast<uintptr_t>(pointer) % alignof(T) == 0;
 }
+} // anon ns
+
+BOOST_FUSION_ADAPT_STRUCT(Segment, (std::uint64_t, from)(std::uint64_t, to))
+BOOST_FUSION_ADAPT_STRUCT(SpeedSource, (unsigned, speed))
+BOOST_FUSION_ADAPT_STRUCT(Turn, (std::uint64_t, from)(std::uint64_t, via)(std::uint64_t, to))
+BOOST_FUSION_ADAPT_STRUCT(PenaltySource, (double, penalty))
+
+// Function parses a list of CSV files using "key,value,comment" grammar.
+// Key and Value structures must be a model of Random Access Sequence.
+// Also the Value structure must have source member that will be filled
+// with the index of the csv_filenames array.
+// Function returns a lambda function that maps input Key to boost::optional<Value>
+template <typename Key, typename Value>
+auto ParseCSV(
+    const std::vector<std::string> &csv_filenames,
+    const qi::rule<boost::spirit::line_pos_iterator<boost::spirit::istream_iterator>, Key()> &key,
+    const qi::rule<boost::spirit::line_pos_iterator<boost::spirit::istream_iterator>, Value()>
```

Comment:
yeah, this is bad looking, because input grammar parameters should know about internals to have correct  `Iterator` type at the compile time. I will move the function into a functor object `CSVParser` and make typedefs `Iterator` and `Rule`, and `operator(const ... &filenames, const Rule& key, const Rule &value)`